Simplified/Reader2/NYPLReaderBookmarksBusinessLogic.swift

```diff
@@ -200,25 +230,113 @@ class NYPLReaderBookmarksBusinessLogic: NSObject, NYPLReadiumViewSyncManagerDele
-//        if !success {
-//          let alert = NYPLAlertUtils.alert(title: "Error Syncing Bookmarks",
-//                                           message: "There was an error syncing bookmarks to the server. Ensure your device is connected to the internet or try again later.")
-//          vc.present(alert, animated: true)
-//        }
-//      }
-//    }
+    
+  func syncBookmarks(completion: @escaping (Bool, [NYPLReadiumBookmark]) -> ()) {
+    NYPLReachability.shared()?.reachability(for: NYPLConfiguration.mainFeedURL(),
+                                            timeoutInternal: 8.0,
+                                            handler: { (reachable) in
+      guard reachable else {
+        self.handleBookmarksSyncFail(level: .warn,
+                                     message: "Error: host was not reachable for bookmark sync attempt.",
+                                     completion: completion)
+        return
+      }
+                    
+      Log.debug(#file, "Syncing bookmarks...")
+      // First check for and upload any local bookmarks that have never been saved to the server.
+      // Wait til that's finished, then download the server's bookmark list and filter out any that can be deleted.
+      let localBookmarks = self.bookRegistryProvider.readiumBookmarks(forIdentifier: self.book.identifier)
+      NYPLAnnotations.uploadLocalBookmarks(localBookmarks, forBook: self.book.identifier) { (bookmarksUploaded, bookmarksFailedToUpload) in
+        for localBookmark in localBookmarks {
+          for uploadedBookmark in bookmarksUploaded {
+            if localBookmark.isEqual(uploadedBookmark) {
+              self.bookRegistryProvider.replace(localBookmark, with: uploadedBookmark, forIdentifier: self.book.identifier)
+            }
+          }
+        }
+        
+        NYPLAnnotations.getServerBookmarks(forBook: self.book.identifier, atURL: self.book.annotationsURL) { (serverBookmarks) in
+          guard let serverBookmarks = serverBookmarks else {
+            self.handleBookmarksSyncFail(level: .debug,
+                                         message: "Ending sync without running completion. Returning original list of bookmarks.",
+                                         completion: completion)
+            return
+          }
+            
+          Log.debug(#file, serverBookmarks.count == 0 ? "No server bookmarks" : "Server bookmarks count: \(serverBookmarks.count)")
+          
+          self.updateLocalBookmarks(serverBookmarks: serverBookmarks,
+                                     localBookmarks: localBookmarks,
+                                     bookmarksFailedToUpload: bookmarksFailedToUpload)
+          { [weak self] in
+            guard let self = self else {
+              completion(false, localBookmarks)
+              return
+            }
+            self.bookmarks = self.bookRegistryProvider.readiumBookmarks(forIdentifier: self.book.identifier)
```

Comment:
since we always use the book registry to get the current bookmarks, do we actually need to have a `bookmarks` property here?

Reply:
The `bookmarks` only being updated when sync/add/delete happens, and the `BookRegistry` does not get updated until the network request is completed, so I think a local property is necessary